File: python/dataproc_templates/gcs/gcs_to_spanner.py
# Copyright 2022 Google LLC
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# https://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

from typing import Dict, Sequence, Optional, Any
from logging import Logger
import argparse
import pprint

from pyspark.sql import SparkSession, DataFrame

from dataproc_templates import BaseTemplate
import dataproc_templates.util.template_constants as constants

__all__ = ['GCSToSpannerTemplate']


class GCSToSpannerTemplate(BaseTemplate):
  """
  Dataproc template implementing loads from GCS into Spanner
  """

  @staticmethod
  def parse_args(args: Optional[Sequence[str]] = None) -> Dict[str, Any]:
    parser: argparse.ArgumentParser = argparse.ArgumentParser()

    parser.add_argument(
      f'--{constants.GCS_SPANNER_INPUT_LOCATION}',
      dest=constants.GCS_SPANNER_INPUT_LOCATION,
      required=True,
      help='GCS location of the input files'
    )
    parser.add_argument(
      f'--{constants.GCS_SPANNER_INPUT_FORMAT}',
      dest=constants.GCS_SPANNER_INPUT_FORMAT,
      required=True,
      # TODO: Add support to CSV?
      help='Input file format (one of: <avro,parquet,orc>)',
      choices=[
        constants.FORMAT_AVRO,
        constants.FORMAT_PRQT,
        constants.FORMAT_ORC,
      ]
    )
    parser.add_argument(
      f'--{constants.GCS_SPANNER_OUTPUT_INSTANCE}',
      dest=constants.GCS_SPANNER_OUTPUT_INSTANCE,
      required=True,
      help='Spanner output instance'
    )
    parser.add_argument(
      f'--{constants.GCS_SPANNER_OUTPUT_DATABASE}',
      dest=constants.GCS_SPANNER_OUTPUT_DATABASE,
      required=True,
      help='Spanner output database'
    )
    parser.add_argument(
      f'--{constants.GCS_SPANNER_OUTPUT_TABLE}',
      dest=constants.GCS_SPANNER_OUTPUT_TABLE,
      required=True,
      help='Spanner output table'
    )
    parser.add_argument(
      f'--{constants.GCS_SPANNER_PRIMARY_KEY}',
      dest=constants.GCS_SPANNER_PRIMARY_KEY,
      required=True,
      help='Primary key of the table'
    )
    parser.add_argument(
      f'--{constants.GCS_SPANNER_OUTPUT_SAVE_MODE}',
      dest=constants.GCS_SPANNER_OUTPUT_SAVE_MODE,
      required=False,
      default=constants.OUTPUT_MODE_APPEND,
      help=(
        'Output write mode '
        '(one of: <Overwrite|ErrorIfExists|Append|Ignore>) '
        '(Defaults to ErrorIfExists)'
      ),
      choices=[
        constants.OUTPUT_MODE_OVERWRITE,
        constants.OUTPUT_MODE_APPEND,
        constants.OUTPUT_MODE_IGNORE,
        constants.OUTPUT_MODE_ERRORIFEXISTS
      ]
    )
    parser.add_argument(
      f'--{constants.GCS_SPANNER_OUTPUT_BATCH_INSERT_SIZE}',
      dest=constants.GCS_SPANNER_OUTPUT_BATCH_INSERT_SIZE,
      required=False,
      default=1000,
      help='Spanner batch insert size '
    )

    known_args: argparse.Namespace
    known_args, _ = parser.parse_known_args(args)

    return vars(known_args)

  def run(self, spark: SparkSession, args: Dict[str, Any]) -> None:

    project_id: str = constants.PROJECT_ID_PROP
    logger: Logger = self.get_logger(spark=spark)

    # Arguments
    input_file_location: str = args[constants.GCS_SPANNER_INPUT_LOCATION]
    input_file_format: str = args[constants.GCS_SPANNER_INPUT_FORMAT]
    spanner_instance: str = args[constants.GCS_SPANNER_OUTPUT_INSTANCE]
    spanner_database: str = args[constants.GCS_SPANNER_OUTPUT_DATABASE]
    spanner_table: str = args[constants.GCS_SPANNER_OUTPUT_TABLE]
    spanner_pk: str = args[constants.GCS_SPANNER_PRIMARY_KEY]
    spanner_save_mode: str = args[constants.GCS_SPANNER_OUTPUT_SAVE_MODE]
    batch_size: str = args[constants.GCS_SPANNER_OUTPUT_BATCH_INSERT_SIZE]
    spanner_url: str = (
      f"jdbc:cloudspanner:/projects/{project_id}/"
      f"instances/{spanner_instance}/"
      f"databases/{spanner_database}"
    )
    spanner_driver = "com.google.cloud.spanner.jdbc.JdbcDriver"
    table_options = f"PRIMARY KEY ({spanner_pk})"

    logger.info(
      "Starting GCS to Spanner spark job with parameters:\n"
      f"{pprint.pformat(args)}"
    )

    # Read
    input_data: DataFrame

    input_data = (
      spark.read
      .format(input_file_format)
      .load(input_file_location)
    )

    # Write
    (
      input_data.write
      .format("jdbc")
      .option("url", spanner_url)
      .option("dbtable", spanner_table)
      .option("createTableOptions", table_options)
      .option("isolationLevel", None)
      .option("batchsize", batch_size)
      .option("driver", spanner_driver)
      .mode(spanner_save_mode)
      .save()
    )

File: java/src/main/java/com/google/cloud/dataproc/templates/gcs/GCSToSpanner.java
/*
 * Copyright (C) 2021 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not
 * use this file except in compliance with the License. You may obtain a copy of
 * the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
 * License for the specific language governing permissions and limitations under
 * the License.
 */
package com.google.cloud.dataproc.templates.gcs;

import static com.google.cloud.dataproc.templates.util.TemplateConstants.SPANNER_GOOGLESQL_JDBC_DIALECT;
import static com.google.cloud.dataproc.templates.util.TemplateConstants.SPANNER_POSTGRESQL_JDBC_DIALECT;

import com.google.cloud.dataproc.dialects.PostgresJDBCDialect;
import com.google.cloud.dataproc.dialects.SpannerJdbcDialect;
import com.google.cloud.dataproc.templates.BaseTemplate;
import com.google.cloud.dataproc.templates.util.PropertyUtil;
import com.google.cloud.dataproc.templates.util.ValidationUtil;
import org.apache.spark.sql.Dataset;
import org.apache.spark.sql.Row;
import org.apache.spark.sql.SaveMode;
import org.apache.spark.sql.SparkSession;
import org.apache.spark.sql.execution.datasources.jdbc.JDBCOptions;
import org.apache.spark.sql.jdbc.JdbcDialects;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

public class GCSToSpanner implements BaseTemplate {

  public static final String SPANNER_JDBC_DRIVER = "com.google.cloud.spanner.jdbc.JdbcDriver";

  private static final Logger LOGGER = LoggerFactory.getLogger(GCSToSpanner.class);
  private final GCSToSpannerConfig config;

  public GCSToSpanner(GCSToSpannerConfig config) {
    this.config = config;
  }

  public static GCSToSpanner of(String... args) {
    GCSToSpannerConfig config = GCSToSpannerConfig.fromProperties(PropertyUtil.getProperties());
    LOGGER.info("Config loaded\n{}", config);
    return new GCSToSpanner(config);
  }

  @Override
  public void runTemplate() {

    try (SparkSession spark = SparkSession.builder().appName("GCS to Spanner").getOrCreate()) {
      // Set log level
      spark.sparkContext().setLogLevel(config.getSparkLogLevel());

      Dataset<Row> dataset =
          spark.read().format(config.getInputFormat()).load(config.getInputLocation());

      try {
        write(dataset);
      } catch (Exception ex) {
        LOGGER.error("Exception in GCSToSpanner Template", ex);
      }
    }
  }

  public void write(Dataset<Row> dataset) throws Exception {
    String spannerUrl =
        String.format(
            "jdbc:cloudspanner:/projects/%s/instances/%s/databases/%s?lenient=true",
            config.getProjectId(), config.getInstance(), config.getDatabase());
    LOGGER.info("Spanner JDBC Dialect is {}", config.getSpannerJdbcDialect());
    switch (config.getSpannerJdbcDialect().toLowerCase()) {
      case SPANNER_GOOGLESQL_JDBC_DIALECT:
        JdbcDialects.registerDialect(new SpannerJdbcDialect());
        break;

      case SPANNER_POSTGRESQL_JDBC_DIALECT:
        JdbcDialects.registerDialect(new PostgresJDBCDialect());
        if (config.getSaveMode() != SaveMode.Append) {
          throw new UnsupportedOperationException(
              "Spanner jdbc dialect supports only append mode. Please refer README.md file.");
        }
        break;

      default:
        throw new UnsupportedOperationException(
            String.format(
                "%s spanner jdbc dialect is not supported. Expected values are either googlesql or postgresql.",
                config.getSpannerJdbcDialect()));
    }
    LOGGER.info("Start writing to spanner");
    dataset
        .write()
        .format("jdbc")
        .option(JDBCOptions.JDBC_URL(), spannerUrl)
        .option(JDBCOptions.JDBC_TABLE_NAME(), config.getTable())
        .option(
            JDBCOptions.JDBC_CREATE_TABLE_OPTIONS(),
            String.format("PRIMARY KEY (%s)", config.getPrimaryKey()))
        .option(
            JDBCOptions.JDBC_TXN_ISOLATION_LEVEL(),
            "NONE") // Needed because transaction have a 20,000 mutation limit per commit.
        .option(
            JDBCOptions.JDBC_BATCH_INSERT_SIZE(), config.getBatchInsertSize()) // default is 1000
        .option(JDBCOptions.JDBC_DRIVER_CLASS(), SPANNER_JDBC_DRIVER)
        .mode(config.getSaveMode())
        .save();
  }

  @Override
  public void validateInput() {
    ValidationUtil.validateOrThrow(config);
  }
}
